demos/sslecho/main.c

```diff
@@ -0,0 +1,330 @@
+/*
+ *  Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ *  Licensed under the Apache License 2.0 (the "License").  You may not use
+ *  this file except in compliance with the License.  You can obtain a copy
+ *  in the file LICENSE in the source distribution or at
+ *  https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+
+static const char *revision = "1.0.7";
+
+static const int server_port = 4433;
+
+typedef unsigned char   bool;
+#define true            1
+#define false           0
+
+/*
+ * This flag won't be useful until both accept/read (TCP & SSL) methods
+ * can be called with a timeout. TBD.
+ */
+static volatile bool    server_running = true;
+
+int create_socket(bool isServer) {
+    int s;
+    int optval = 1;
+    struct sockaddr_in addr;
+
+    s = socket(AF_INET, SOCK_STREAM, 0);
+    if (s < 0) {
+        perror("Unable to create socket");
+        exit(EXIT_FAILURE);
+    }
+
+    if (isServer) {
+        addr.sin_family = AF_INET;
+        addr.sin_port = htons(server_port);
+        addr.sin_addr.s_addr = INADDR_ANY;
+
+        /* Reuse the address; good for quick restarts */
+        if (setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval))
+                < 0) {
+            perror("setsockopt(SO_REUSEADDR) failed");
+            exit(EXIT_FAILURE);
+        }
+
+        if (bind(s, (struct sockaddr*) &addr, sizeof(addr)) < 0) {
+            perror("Unable to bind");
+            exit(EXIT_FAILURE);
+        }
+
+        if (listen(s, 1) < 0) {
+            perror("Unable to listen");
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    return s;
+}
+
+SSL_CTX* create_context(bool isServer) {
+    const SSL_METHOD *method;
+    SSL_CTX *ctx;
+
+    if (isServer)
+        method = TLS_server_method();
+    else
+        method = TLS_client_method();
+
+    ctx = SSL_CTX_new(method);
+    if (ctx == NULL) {
+        perror("Unable to create SSL context");
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+
+    return ctx;
+}
+
+void configure_server_context(SSL_CTX *ctx) {
+    /* Set the key and cert */
+    if (SSL_CTX_use_certificate_file(ctx, "cert.pem", SSL_FILETYPE_PEM) <= 0) {
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+
+    if (SSL_CTX_use_PrivateKey_file(ctx, "key.pem", SSL_FILETYPE_PEM) <= 0) {
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+}
+
+void configure_client_context(SSL_CTX *ctx) {
+    /*
+     * Configure the client to abort the handshake if certificate verification
+     * fails
+     */
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
+    if (!SSL_CTX_load_verify_file(ctx, "cert.pem")) {
```

Comment:
Not understanding what's requested here...
Seems that you're proposing two modifications; one with 'set_verify', and one with 'load_verify'. One or both?
Would appreciate clarification. Perhaps show how the code should be written.

Reply:
It's an either/or scenario. Pick one or the other of @vdukhovni's suggestions. We are cheating a little at the moment, because we are using a self-signed certificate and therefore the client needs to trust the certificate itself directly. So the current code says use the "cert.pem" file directly as our trust store of trusted certificates. More commonly you would have a file full of CA certificates that we trust, or a directory full of them.

Actually calling `SSL_CTX_load_verify_locations` and specifying a `CAfile` and a NULL `CApath` is precisely equivalent to just calling `SSL_CTX_load_verify_file`.

I'm not really sure why @vdukhovni prefers `SSL_CTX_load_verify_locations` over `SSL_CTX_load_verify_file` since they do the same thing in the case where the `CApath` parameter is NULL (in fact the former is implemented in terms of the latter) - although in a *real* scenario you wouldn't be using a self-signed certificate and you would have multiple CA certs.

Using the default system certificate store is probably want most *real* applications will want to do - but doesn't work so well for this demo since we are using a self-signed cert.

Reply:
> Using the default system certificate store is probably want most _real_ applications will want to do - but doesn't work so well for this demo since we are using a self-signed cert.

I'd add a comment to clarify this.

Reply:
Yes, for this to be closer to code that a real application might use with modest changes, it should use the system certificate store, and perhaps support `CApath` as well.  Here, you can't just use the default locations (well you can by setting `SSL_CERT_FILE` or `SSL_CERT_DIR` in the process environment).

The idea is to do the common thing as much as possible.